File: wagerproof-ios-native/Wagerproof/Features/Onboarding/PostOnboardingPaywall.swift
// PostOnboardingPaywall.swift
//
// Host for the post-onboarding paywall. Mounts above the main app shell
// once onboarding finishes and the user is not yet a Pro subscriber.
//
// Renderer is remote-configured via the placement offering's metadata
// (dashboard-editable, no app release):
//   - `custom_paywall_enabled` (default true): the fully custom SwiftUI
//     `CustomPaywallView` — RevenueCat as data layer only. `false` is the
//     kill switch back to the legacy `RevenueCatUI.PaywallView` template.
//   - `paywall_close_enabled` (default FALSE = hard): the real onboarding gate
//     ships HARD — no X on either renderer. Set it `true` to soften the gate
//     (e.g. App Review builds). The error/timeout "Continue without
//     subscription" escape survives BOTH modes.
//
// The Secret-Settings debug preview (`isDebugPreview: true`) overrides the above
// and always shows a bright red DEBUG close button, so a tester can escape the
// otherwise-non-dismissible gate and knows the run is a debug invocation.
//
// The host owns:
//   - Gating predicate (auth + onboarding complete + !isPro + !dismissed —
//     evaluated by RootView's `shouldPresentPaywall`).
//   - Placement offering fetch + retry / skip fallbacks if RC is unreachable.
//   - Post-purchase finalization: refresh `RevenueCatStore` so the rest of
//     the app re-renders with the granted entitlement, then dismiss.
//   - Meta analytics fan-out on conversion (attribution-critical path).
//     Mixpanel paywall funnel events fire inside `CustomPaywallView`.

import SwiftUI
import RevenueCat
import RevenueCatUI
import WagerproofDesign
import WagerproofServices
import WagerproofStores

struct PostOnboardingPaywall: View {
    @Environment(AuthStore.self) private var authStore
    @Environment(OnboardingStore.self) private var onboarding
    @Environment(RevenueCatStore.self) private var revenueCat
    @Environment(ProAccessStore.self) private var proAccess

    /// Tells the host (`RootView`) the user is done with the paywall — fires
    /// after a successful purchase/restore, when the dashboard-configured
    /// close button is tapped, OR when the skip fallback is used. The host
    /// flips its own `paywallDismissed` flag, which drops the
    /// `fullScreenCover` binding to `false` and animates the cover away.
    /// Lifting this state out of the paywall avoids the blank-modal trap
    /// where flipping a local `dismissed` flag would render `EmptyView()`
    /// behind a cover the host doesn't know to close.
    let onUserDismissed: () -> Void

    /// True only when presented from Secret Settings for testing. Forces the
    /// paywall closeable via a red DEBUG button so the real (hard) onboarding
    /// gate stays non-dismissible while a tester can still escape the preview.
    var isDebugPreview: Bool = false

    @State private var offering: Offering?
    @State private var isLoadingOffering: Bool = true
    @State private var loadError: String?
    /// True while we're awaiting the post-purchase `refreshCustomerInfo()`.
    /// Forces a spinner over the paywall surface so the user doesn't see a
    /// flash of "free" before the entitlement applies.
    @State private var isFinalizing: Bool = false
    /// Flips after 10s if the offering fetch hasn't resolved — surfaces the
    /// retry/skip affordance so the user is never stuck on a spinner.
    @State private var timedOut: Bool = false

    var body: some View {
        paywallSurface
            .preferredColorScheme(.dark)
            .task { await loadOffering() }
            .task { await startTimeoutWatchdog() }
            // Block the swipe-to-dismiss gesture on the cover sheet so the
            // paywall stays "non-dismissible" unless the user purchases,
            // restores, or hits the dashboard X. Matches RN's
            // `onRequestClose={() => {}}` Modal behavior.
            .interactiveDismissDisabled(true)
    }

    /// Remote hard/soft switch (offering metadata `paywall_close_enabled`).
    /// Metadata booleans arrive as NSNumber from the dashboard JSON — the
    /// `as? Bool` bridge handles true/false and 0/1. Absent key = HARD (no X);
    /// the real onboarding gate ships hard, a remote `true` softens it. The
    /// debug preview always allows closing (via the red DEBUG button).
    private var closeEnabled: Bool {
        if isDebugPreview { return true }
        return (offering?.metadata["paywall_close_enabled"] as? Bool) ?? false
    }

    /// Remote renderer switch (offering metadata `custom_paywall_enabled`).
    /// Absent key = custom paywall; explicit `false` = legacy RC template.
    private var customPaywallEnabled: Bool {
        (offering?.metadata["custom_paywall_enabled"] as? Bool) ?? true
    }

    private var isShowingCustomPaywall: Bool {
        customPaywallEnabled && offering != nil && !isLoadingOffering && loadError == nil
    }

    @ViewBuilder
    private var paywallSurface: some View {
        ZStack {
            Color.black.ignoresSafeArea()

            if let offering, !isLoadingOffering, loadError == nil {
                if customPaywallEnabled {
                    CustomPaywallView(
                        offering: offering,
                        allowClose: closeEnabled,
                        source: "post_onboarding",
                        // Monetization stays on the core WagerProof green so
                        // plan selection and purchase always read as the same
                        // product, regardless of the temporary onboarding tint.
                        accent: .appPrimary,
                        agentName: onboarding.agentDraft.name,
                        spriteIndex: onboarding.agentDraft.spriteIndex ?? 0,
                        researchBucketRaw: onboarding.survey.researchTimeBucket,
                        onPurchaseFinalized: { transaction, customerInfo in
                            Task { await finalize(transaction: transaction, customerInfo: customerInfo) }
                        },
                        onRequestClose: onUserDismissed,
                        debugClose: isDebugPreview
                    )
                } else {
                    // Legacy dashboard-owned template (kill-switch path).
                    // `displayCloseButton` only renders on V1 templates.
                    PaywallView(offering: offering, displayCloseButton: closeEnabled)
                        .onPurchaseCompleted { transaction, customerInfo in
                            Task { await finalize(transaction: transaction, customerInfo: customerInfo) }
                        }
                        .onRestoreCompleted { customerInfo in
                            // Restore can complete WITHOUT a granted entitlement
                            // (e.g. user restored on a device with no purchase
                            // history). Only collapse the paywall when the WagerProof
                            // Pro entitlement is actually active. Matches the RN
                            // guard `customerInfo?.entitlements?.active?.['WagerProof Pro']`.
                            guard customerInfo.entitlements.active[RevenueCatService.entitlementIdentifier] != nil else {
                                return
                            }
                            Task { await finalize(transaction: nil, customerInfo: customerInfo) }
                        }
                        .onRequestedDismissal {
                            // Dashboard-configured close button — explicit user
                            // intent to bail. Let them through.
                            onUserDismissed()
                        }
                }
            }

            if isLoadingOffering || isFinalizing {
                loadingOverlay
            } else if loadError != nil || offering == nil {
                errorOverlay
            }

            // Guaranteed X overlay for the LEGACY branch (V2 RC templates
            // ignore `displayCloseButton` and use whatever the dashboard
            // configured) and for the loading state. The custom paywall
            // draws its own metadata-gated X, so the overlay would double
            // up there. Hidden entirely in hard mode — the error surface's
            // "Continue without subscription" remains the escape hatch.
            if closeEnabled && !isShowingCustomPaywall {
                closeOverlay
            }
        }
    }

    private var closeOverlay: some View {
        VStack {
            HStack {
                Spacer()
                Button(action: { onUserDismissed() }) {
                    if isDebugPreview {
                        // Matches CustomPaywallView.debugCloseLabel so the escape
                        // hatch reads the same whether the custom paywall or the
                        // loading/legacy surface is on screen.
                        HStack(spacing: 5) {
                            Image(systemName: "xmark")
                                .font(.system(size: 12, weight: .heavy))
                            Text("DEBUG")
                                .font(.system(size: 11, weight: .heavy, design: .monospaced))
                                .tracking(0.5)
                        }
                        .foregroundStyle(.white)
                        .padding(.horizontal, 12)
                        .frame(height: 34)
                        .background(Capsule().fill(Color.red))
                        .overlay(Capsule().strokeBorder(Color.white.opacity(0.55), lineWidth: 1))
                    } else {
                        Image(systemName: "xmark")
                            .font(.system(size: 14, weight: .bold))
                            .foregroundStyle(.white)
                            .frame(width: 32, height: 32)
                            .background(
                                Circle().fill(Color.black.opacity(0.55))
                            )
                            .overlay(
                                Circle().strokeBorder(Color.white.opacity(0.18), lineWidth: 1)
                            )
                    }
                }
                .padding(.trailing, 16)
                .padding(.top, 8)
                .accessibilityLabel(isDebugPreview ? "Close debug paywall" : "Close paywall")
            }
            Spacer()
        }
    }

    // MARK: - Overlays

    private var loadingOverlay: some View {
        VStack(spacing: Spacing.md) {
            ProgressView()
                .progressViewStyle(.circular)
                .tint(Color.appPrimary)
                .scaleEffect(1.4)

            Text(isFinalizing ? "Finalizing your subscription..." : "Loading subscription options...")
                .font(.system(size: 16))
                .foregroundStyle(Color.white.opacity(0.75))
                .multilineTextAlignment(.center)
                .padding(.horizontal, 32)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Color.black.ignoresSafeArea())
    }

    private var errorOverlay: some View {
        VStack(spacing: 24) {
            Spacer()

            Image(systemName: "exclamationmark.triangle.fill")
                .font(.system(size: 56, weight: .semibold))
                .foregroundStyle(Color.appPrimary)

            VStack(spacing: 12) {
                Text("Unable to load subscription options")
                    .font(.system(size: 22, weight: .bold))
                    .foregroundStyle(.white)
                    .multilineTextAlignment(.center)

                if let loadError {
                    Text(loadError)
                        .font(.system(size: 15))
                        .foregroundStyle(Color.white.opacity(0.7))
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 32)
                }
            }

            Spacer()

            VStack(spacing: 12) {
                OnboardingLiquidGlassButton(title: "Retry") {
                    Task { await loadOffering() }
                }
                OnboardingLiquidGlassButton(
                    title: "Continue without subscription",
                    tint: Color.white.opacity(0.18)
                ) {
                    // Same escape hatch as the RN "skip" button — keeps the
                    // user moving when the network is down so onboarding
                    // can't strand them on a dead paywall.
                    onUserDismissed()
                }
            }
            .padding(.horizontal, 24)
            .padding(.bottom, 40)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Color.black.ignoresSafeArea())
    }

    // MARK: - Logic

    private func loadOffering() async {
        isLoadingOffering = true
        loadError = nil
        timedOut = false

        // Prefer the placement-specific offering so the dashboard can ship a
        // distinct post-onboarding variant; the service helper falls back to
        // the current offering when no placement is attached.
        if let placementOffering = await revenueCat.fetchOffering(forPlacement: RevenueCatService.Placement.onboarding) {
            offering = placementOffering
            isLoadingOffering = false
            return
        }

        // Last-resort fallback — use whatever offering RevenueCatStore already
        // cached from `refreshOffering()`. Keeps the paywall functional even
        // if the placement returns nil and the live fetch fails.
        if let fallback = revenueCat.offering {
            offering = fallback
            isLoadingOffering = false
            return
        }

        offering = nil
        loadError = "Couldn't reach the subscription service. Check your connection and try again."
        isLoadingOffering = false
    }

    /// Safety watchdog — if the offering fetch hasn't finished within 10s,
    /// flip into the error/retry surface so the user isn't trapped behind a
    /// spinner. Matches the RN `setTimeout(() => setTimedOut(true), 10000)`.
    private func startTimeoutWatchdog() async {
        try? await Task.sleep(nanoseconds: 10_000_000_000)
        guard !Task.isCancelled else { return }
        guard isLoadingOffering && offering == nil else { return }
        timedOut = true
        isLoadingOffering = false
        if loadError == nil {
            loadError = "Subscription options are taking longer than expected to load."
        }
    }

    private func finalize(transaction: StoreTransaction?, customerInfo: CustomerInfo) async {
        isFinalizing = true
        // Trusted refresh from RC servers — `RevenueCatStore.refreshCustomerInfo`
        // is allowed to flip granted/denied (the trust-downgrade guard only
        // protects against stream updates), so the rest of the app sees the
        // new entitlement immediately after this returns.
        await revenueCat.refreshCustomerInfo()

        // Fan out Meta analytics so ad-network LTV / Subscribe events fire
        // before the user is dropped into the main app. Best-effort — never
        // blocks the dismissal.
        if let transaction {
            trackMetaConversion(transaction: transaction, customerInfo: customerInfo)
        }

        isFinalizing = false
        onUserDismissed()
    }

    /// Fire Meta SDK Subscribe / Purchase events using the price + currency
    /// extracted from the matched RevenueCat package. Mirrors the RN mapping
    /// in `services/analytics.ts` `trackFacebookSubscriptionEvent` where
    /// trials map to `fb_mobile_purchase` and paid subs map to `Subscribe`.
    private func trackMetaConversion(transaction: StoreTransaction, customerInfo: CustomerInfo) {
        // Resolve the purchased package off the offering so we can pull
        // price + currency off the StoreProduct. StoreTransaction itself
        // doesn't expose the price (RC keeps it on the catalog side).
        guard let offering,
              let package = offering.availablePackages.first(where: { $0.storeProduct.productIdentifier == transaction.productIdentifier }) else {
            return
        }

        let product = package.storeProduct
        let amount = product.price
        let currency = product.currencyCode ?? "USD"

        // Subscription type derived from the active entitlement so the
        // dashboards roll up correctly under monthly / yearly / lifetime.
        let entitlement = customerInfo.entitlements.active[RevenueCatService.entitlementIdentifier]
        let productId = (entitlement?.productIdentifier ?? transaction.productIdentifier).lowercased()
        let subscriptionType: String = {
            if productId.contains("lifetime") { return "lifetime" }
            if productId.contains("annual") || productId.contains("yearly") { return "yearly" }
            if productId.contains("monthly") { return "monthly" }
            return "unknown"
        }()

        let contentId = "\(subscriptionType)_subscription"
        // RN computes a coarse LTV multiplier so Meta's optimization model
        // gets a forward-looking value rather than just the first month.
        let predictedLtv: Decimal = {
            switch subscriptionType {
            case "monthly": return amount * 4
            case "yearly": return amount * (Decimal(string: "1.3") ?? 1)
            default: return amount
            }
        }()

        let isTrial = entitlement?.periodType == .trial

        let metaParameters: [String: Any] = [
            "fb_currency": currency,
            "fb_content_type": "product",
            "fb_content_id": contentId,
            "fb_order_id": transaction.transactionIdentifier,
            "fb_predicted_ltv": NSDecimalNumber(decimal: predictedLtv).stringValue,
            "fb_success": "1",
            "fb_payment_info_available": "1",
        ]

        if isTrial {
            // Trial start → `fb_mobile_purchase` so Meta sees attribution at
            // the same point as the RevenueCat server-side mapping.
            MetaAnalyticsService.shared.trackPurchase(
                amount: amount,
                currency: currency,
                parameters: metaParameters
            )
        } else {
            // Paid first sub → `Subscribe` w/ valueToSum so revenue dashboards
            // attribute the LTV to the install cohort.
            MetaAnalyticsService.shared.trackSubscribe(
                amount: amount,
                currency: currency,
                parameters: metaParameters
            )
        }

        // Force-flush so the event hits Meta's pipeline before the user
        // backgrounds the app post-purchase.
        MetaAnalyticsService.shared.flush()
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/paywall/PostOnboardingPaywall.kt
package com.wagerproof.app.features.paywall

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.WindowInsetsSides
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.only
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeDrawing
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.windowInsetsPadding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.revenuecat.purchases.CustomerInfo
import com.revenuecat.purchases.Offering
import com.revenuecat.purchases.Package
import com.revenuecat.purchases.models.StoreTransaction
import com.revenuecat.purchases.ui.revenuecatui.Paywall
import com.revenuecat.purchases.ui.revenuecatui.PaywallListener
import com.revenuecat.purchases.ui.revenuecatui.PaywallOptions
import com.wagerproof.app.di.appGraph
import com.wagerproof.core.design.icons.AppIcon
import com.wagerproof.core.design.tokens.AppColors
import com.wagerproof.core.design.tokens.AppTypography
import com.wagerproof.core.design.tokens.Spacing
import com.wagerproof.core.services.AnalyticsService
import com.wagerproof.core.services.PaywallConversionTracker
import com.wagerproof.core.services.RevenueCatService
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/** Attribution placement id for this surface — matches iOS's `"post_onboarding"`. */
private const val PAYWALL_SOURCE = "post_onboarding"

/**
 * Host for the post-onboarding paywall — port of iOS `PostOnboardingPaywall.swift`.
 *
 * Mounted by [com.wagerproof.app.nav.RootHost] above the main shell once
 * onboarding finishes and the user isn't yet Pro.
 *
 * The renderer and the gate hardness are remote-configured from the placement
 * offering's metadata (dashboard-editable, no app release):
 *  - `custom_paywall_enabled` (default TRUE): the fully custom [CustomPaywallView]
 *    with RevenueCat as data layer only. `false` is the kill switch back to the
 *    RevenueCatUI template.
 *  - `paywall_close_enabled` (default FALSE = HARD): the real onboarding gate
 *    ships hard — no ✕ on either renderer, and back is swallowed. Flip it `true`
 *    to soften the gate (e.g. an App Review build). The error/timeout
 *    "Continue without subscription" escape survives BOTH modes, so a user can
 *    never be stranded on a paywall that failed to load.
 *
 * [isDebugPreview] (Secret Settings) overrides both and always draws a bright red
 * DEBUG close, so a tester can escape the otherwise-non-dismissible gate and can
 * see at a glance that the run is a debug invocation.
 *
 * The host owns the offering fetch + retry/skip fallbacks and post-purchase
 * finalization (refresh [com.wagerproof.core.stores.RevenueCatStore], fan out
 * attribution, dismiss). Mixpanel paywall-funnel events fire inside
 * [CustomPaywallView] on the custom path.
 */
@Composable
fun PostOnboardingPaywall(
    onUserDismissed: () -> Unit,
    isDebugPreview: Boolean = false,
    /**
     * Reports the resolved `paywall_close_enabled` state to the host so the
     * modal's back handler and the ✕ stay in step. Without this the hard gate
     * would only be half applied: no ✕, but back still dismissing.
     */
    onCloseEnabledChanged: (Boolean) -> Unit = {},
) {
    val graph = appGraph()
    val revenueCat = graph.revenueCat
    val onboarding = graph.onboarding
    val scope = rememberCoroutineScope()

    var offering by remember { mutableStateOf<Offering?>(null) }
    var isLoadingOffering by remember { mutableStateOf(true) }
    var loadError by remember { mutableStateOf<String?>(null) }
    var isFinalizing by remember { mutableStateOf(false) }
    var reloadKey by remember { mutableIntStateOf(0) }

    suspend fun loadOffering() {
        isLoadingOffering = true
        loadError = null

        // Prefer the placement-specific offering so the dashboard can ship a
        // distinct post-onboarding variant; fall back to the cached offering.
        val placementOffering = revenueCat.fetchOffering(RevenueCatService.Placement.ONBOARDING)
        if (placementOffering != null) {
            offering = placementOffering
            isLoadingOffering = false
            return
        }
        val fallback = revenueCat.offering
        if (fallback != null) {
            offering = fallback
            isLoadingOffering = false
            return
        }
        offering = null
        loadError = "Couldn't reach the subscription service. Check your connection and try again."
        isLoadingOffering = false
    }

    LaunchedEffect(reloadKey) { loadOffering() }

    val current = offering
    val metadata = current?.metadata.orEmpty()

    // Absent key = HARD (no ✕): the real onboarding gate ships hard and a remote
    // `true` softens it. The debug preview always allows closing.
    val closeEnabled = isDebugPreview ||
        PaywallPlanResolver.metadataBoolean(metadata, "paywall_close_enabled", default = false)

    // Absent key = custom paywall; explicit `false` = legacy RC template.
    val customPaywallEnabled =
        PaywallPlanResolver.metadataBoolean(metadata, "custom_paywall_enabled", default = true)

    val isShowingCustomPaywall =
        customPaywallEnabled && current != null && !isLoadingOffering && loadError == null

    LaunchedEffect(closeEnabled) { onCloseEnabledChanged(closeEnabled) }

    // Impression for the LEGACY template only — the custom paywall fires its own
    // `paywall_presented` (with the plan it actually highlighted) from inside
    // CustomPaywallView, and two senders would double-count the funnel step.
    var didTrackPresented by remember { mutableStateOf(false) }
    LaunchedEffect(current, isLoadingOffering, isShowingCustomPaywall) {
        if (didTrackPresented || isLoadingOffering || current == null || isShowingCustomPaywall) {
            return@LaunchedEffect
        }
        didTrackPresented = true
        AnalyticsService.track(
            "paywall_presented",
            mapOf("source" to PAYWALL_SOURCE, "variant" to "revenuecat_template"),
        )
        PaywallConversionTracker.trackPaywallView(source = PAYWALL_SOURCE, offering = current)
    }

    // Safety watchdog — if the fetch hasn't resolved within 10s, flip into the
    // error/retry surface so the user is never trapped behind a spinner.
    LaunchedEffect(reloadKey) {
        delay(10_000)
        if (isLoadingOffering && offering == null) {
            isLoadingOffering = false
            if (loadError == null) {
                loadError = "Subscription options are taking longer than expected to load."
            }
        }
    }

    // Trusted refresh (may flip granted/denied) so the rest of the app sees the
    // new entitlement immediately, then fire attribution, then dismiss.
    fun finalize(transaction: StoreTransaction?, customerInfo: CustomerInfo) {
        scope.launch {
            isFinalizing = true
            revenueCat.refreshCustomerInfo()
            // Deduped by order id inside the tracker, so a billing-listener
            // catch-up — or CustomPaywallView having already reported the same
            // purchase — is a complete no-op. A restore proves access but is not a
            // new conversion: historical entitlements must never synthesize a
            // Subscribe/StartTrial event on reinstall or account recovery.
            if (transaction != null) {
                PaywallConversionTracker.trackConversion(
                    source = PAYWALL_SOURCE,
                    transaction = transaction,
                    customerInfo = customerInfo,
                    pkg = null,
                    offering = offering,
                )
            }
            isFinalizing = false
            onUserDismissed()
        }
    }

    // Forced dark — the app is dark-only, but pin black regardless of template.
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(Color.Black),
    ) {
        if (current != null && !isLoadingOffering && loadError == null) {
            if (customPaywallEnabled) {
                CustomPaywallView(
                    offering = current,
                    allowClose = closeEnabled,
                    source = PAYWALL_SOURCE,
                    // Monetization stays on the core WagerProof green so plan
                    // selection and purchase always read as the same product.
                    accent = AppColors.appPrimary,
                    agentName = onboarding.agentDraft.name,
                    spriteIndex = onboarding.agentDraft.spriteIndex ?: 0,
                    researchBucketRaw = onboarding.survey.researchTimeBucket,
                    stakesBucketRaw = onboarding.survey.weeklyStakesBucket,
                    onPurchaseFinalized = { transaction, customerInfo -> finalize(transaction, customerInfo) },
                    onRequestClose = onUserDismissed,
                    debugClose = isDebugPreview,
                )
            } else {
                // Legacy dashboard-owned template (kill-switch path).
                val options = remember(current, closeEnabled) {
                    // RECONCILE: verify against RevenueCatUI 9.7.0 — PaywallOptions
                    // builder + PaywallListener signatures.
                    PaywallOptions.Builder(
                        dismissRequest = {
                            // In hard mode the template's own dismissal request is
                            // ignored: the only ways out are purchase, restore, or
                            // the error surface's skip.
                            if (closeEnabled) onUserDismissed()
                        },
                    )
                        .setOffering(current)
                        // false, NOT true: [CloseOverlay] below already draws an ✕ in
                        // the same corner and is the only one guaranteed to exist (V2
                        // templates ignore this flag). With it true, any template that
                        // DID honour the flag stacked a second ✕ on top of ours.
                        .setShouldDisplayDismissButton(false)
                        .setListener(object : PaywallListener {
                            override fun onPurchaseStarted(rcPackage: Package) {
                                AnalyticsService.track(
                                    "paywall_checkout_started",
                                    mapOf(
                                        "source" to PAYWALL_SOURCE,
                                        "variant" to "revenuecat_template",
                                        "product_id" to rcPackage.product.id,
                                    ),
                                )
                                // Fires before Play Billing resolves so an abandoned
                                // billing sheet still counts as checkout intent.
                                PaywallConversionTracker.trackCheckoutStarted(
                                    source = PAYWALL_SOURCE,
                                    pkg = rcPackage,
                                )
                            }

                            override fun onPurchaseCompleted(
                                customerInfo: CustomerInfo,
                                storeTransaction: StoreTransaction,
                            ) {
                                AnalyticsService.track(
                                    "paywall_converted",
                                    mapOf("source" to PAYWALL_SOURCE, "variant" to "revenuecat_template"),
                                )
                                finalize(storeTransaction, customerInfo)
                            }

                            override fun onRestoreCompleted(customerInfo: CustomerInfo) {
                                // Restore can complete WITHOUT a granted entitlement
                                // (restored on a device with no purchase history).
                                // Only collapse when "WagerProof Pro" is actually active.
                                if (customerInfo.entitlements.active[RevenueCatService.ENTITLEMENT_IDENTIFIER] == null) {
                                    return
                                }
                                AnalyticsService.track(
                                    "paywall_restore_completed",
                                    mapOf("source" to PAYWALL_SOURCE),
                                )
                                finalize(null, customerInfo)
                            }
                        })
                        .build()
                }
                Paywall(options)
            }
        }

        if (isLoadingOffering || isFinalizing) {
            LoadingOverlay(isFinalizing = isFinalizing)
        } else if (loadError != null || offering == null) {
            ErrorOverlay(
                message = loadError,
                onRetry = { reloadKey++ },
                onSkip = onUserDismissed,
            )
        }

        // ✕ overlay for the LEGACY branch and the loading state only — the custom
        // paywall draws its own metadata-gated close in its top bar, so drawing
        // one here too would double it up. Hidden entirely in hard mode; the
        // error surface's "Continue without subscription" stays the escape hatch.
        if (closeEnabled && !isShowingCustomPaywall) {
            CloseOverlay(isDebugPreview = isDebugPreview, onClose = onUserDismissed)
        }
    }
}

@Composable
private fun LoadingOverlay(isFinalizing: Boolean) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Color.Black),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        CircularProgressIndicator(color = AppColors.appPrimary)
        Text(
            text = if (isFinalizing) "Finalizing your subscription..." else "Loading subscription options...",
            fontSize = 16.sp,
            color = Color.White.copy(alpha = 0.75f),
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = Spacing.md, start = Spacing.xxl, end = Spacing.xxl),
        )
    }
}

@Composable
private fun ErrorOverlay(
    message: String?,
    onRetry: () -> Unit,
    onSkip: () -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Color.Black)
            .navigationBarsPadding()
            .padding(Spacing.xl),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Spacer(Modifier.weight(1f))
        Icon(
            imageVector = AppIcon.EXCLAMATION_TRIANGLE_FILL.imageVector,
            contentDescription = null,
            tint = AppColors.appPrimary,
            modifier = Modifier.size(56.dp),
        )
        Text(
            text = "Unable to load subscription options",
            fontSize = 22.sp,
            fontWeight = FontWeight.Bold,
            color = Color.White,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = Spacing.xl),
        )
        if (message != null) {
            Text(
                text = message,
                fontSize = 15.sp,
                color = Color.White.copy(alpha = 0.7f),
                textAlign = TextAlign.Center,
                modifier = Modifier.padding(top = Spacing.md, start = Spacing.xxl, end = Spacing.xxl),
            )
        }
        Spacer(Modifier.weight(1f))
        GlassButton(title = "Retry", onClick = onRetry)
        Spacer(Modifier.size(Spacing.md))
        // UNCONDITIONAL escape hatch — survives the hard gate. Keeps the user
        // moving when the network is down so onboarding can't strand them on a
        // dead paywall.
        GlassButton(
            title = "Continue without subscription",
            tint = Color.White.copy(alpha = 0.18f),
            onClick = onSkip,
        )
        Spacer(Modifier.size(Spacing.xxl))
    }
}

/** Minimal stand-in for iOS `OnboardingLiquidGlassButton`. */
@Composable
private fun GlassButton(
    title: String,
    tint: Color = AppColors.appPrimary,
    onClick: () -> Unit,
) {
    Text(
        text = title,
        fontSize = 16.sp,
        fontWeight = FontWeight.SemiBold,
        color = Color.White,
        textAlign = TextAlign.Center,
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .background(tint, RoundedCornerShape(14.dp))
            .padding(vertical = Spacing.lg),
    )
}

@Composable
private fun CloseOverlay(isDebugPreview: Boolean, onClose: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .windowInsetsPadding(WindowInsets.safeDrawing.only(WindowInsetsSides.Top + WindowInsetsSides.Horizontal))
            .padding(top = Spacing.sm, end = Spacing.lg),
        horizontalArrangement = Arrangement.End,
    ) {
        if (isDebugPreview) {
            // Matches CustomPaywallView's DEBUG pill so the escape hatch reads the
            // same whether the custom paywall or the loading/legacy surface is up.
            Row(
                modifier = Modifier
                    .height(34.dp)
                    .background(Color.Red, CircleShape)
                    .border(1.dp, Color.White.copy(alpha = 0.55f), CircleShape)
                    .clickable(onClick = onClose)
                    .padding(horizontal = 12.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(5.dp),
            ) {
                Icon(
                    imageVector = AppIcon.XMARK.imageVector,
                    contentDescription = "Close debug paywall",
                    tint = Color.White,
                    modifier = Modifier.size(12.dp),
                )
                Text(
                    text = "DEBUG",
                    style = AppTypography.monoCaption.copy(fontSize = 11.sp),
                    fontWeight = FontWeight.Black,
                    color = Color.White,
                )
            }
        } else {
            Box(
                modifier = Modifier
                    .size(32.dp)
                    .background(Color.Black.copy(alpha = 0.55f), CircleShape)
                    .border(1.dp, Color.White.copy(alpha = 0.18f), CircleShape)
                    .clickable(onClick = onClose),
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    imageVector = AppIcon.XMARK.imageVector,
                    contentDescription = "Close paywall",
                    tint = Color.White,
                )
            }
        }
    }
}
